### ISSUE TITLE
Handle not in git repository without error 500 to avoid confusion

### ISSUE THREAD

**Issue by jeanmonet** [jupyterlab/jupyterlab-git#842]
## Description

Hi. git extension seems to output a rather ugly error message in the bash console when the current working directory is not a git repository, or when switching to such a directory:

With debug mode on, the error seems to be triggered by the following:

```
[D 20:39:53.173 LabApp.jupyterlab_git] Execute ['git', 'rev-parse', '--show-toplevel'] in /opt/files/jyusr.
[D 20:39:53.187 LabApp.jupyterlab_git] Code: 128
    Output: 
    Error: fatal: not a git repository (or any of the parent directories): .git
```

which is followed by the following (showing in the console even without debug mode):

```
[E 20:39:53.187 LabApp] {
      "Host": "127.0.0.1:8888",
      "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:83.0) Gecko/20100101 Firefox/83.0",
      "Accept": "*/*",
      "Accept-Language": "en-US,en;q=0.5",
      "Accept-Encoding": "gzip, deflate, br",
      "Referer": "https://127.0.0.1:8888/lab?",
      "Content-Type": "text/plain;charset=UTF-8",
      "X-Xsrftoken": "2|0d88b5b9|xxx|1607891023",
      "Origin": "https://127.0.0.1:8888",
      "Content-Length": "24",
      "Connection": "keep-alive",
      "Cookie": "username-127-0-0-1-8888=\"2|1:0|10:1607891028|23:username-127-0-0-1-8888|44:xxx=|xxxx\"; _xsrf=2|0d88b5b9|xxx|1607891023",
      "Pragma": "no-cache",
      "Cache-Control": "no-cache"
    }
[E 20:39:53.188 LabApp] 500 POST /git/show_top_level?1607891993159 (127.0.0.1) 23.65ms referer=https://127.0.0.1:8888/lab?
```

Am I doing anything wrong or is this expected behavior?

## Reproduce

Launch Jupyter Lab. Working directory is not a git repository.

## Expected behavior

Should not show anything even if I'm not in a git repo directory.

## Context

```
jupyter                   1.0.0                      py_2    conda-forge
jupyter-lsp               0.9.3                    pypi_0    pypi
jupyter_client            6.1.7                      py_0    conda-forge
jupyter_console           6.2.0                      py_0    conda-forge
jupyter_core              4.7.0            py38h578d9bd_0    conda-forge
jupyterlab                2.2.9                      py_0    conda-forge
jupyterlab-git            0.23.2             pyhd8ed1ab_0    conda-forge
jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge
jupyterlab_server         1.2.0                      py_0    conda-forge

gitpython                 3.1.11                     py_0    conda-forge
ipython                   7.19.0           py38h81c977d_0    conda-forge
python                    3.8.6           h852b56e_0_cpython    conda-forge

jupyterlab-git            0.23.2             pyhd8ed1ab_0    conda-forge
```

- Operating System and its version: Ubuntu 20.04

**Comment by fcollonval** [comment]
@jeanmonet thank you for reporting. This is the expected behavior. The extension asks the server if the current folder is a git repository. If it is, an valid response is returned. Otherwise an error is emitted. In such case the front-end handles it to hide the git panel and suggests actions to create a git repository.

**Comment by jeanmonet** [comment]
Hi thanks for the reply. I think I understood the logic, but what surprised me is that a 500 error appears in the terminal (without debug mode), with an output of the request that caused the 500 error. Is this also part of normal behavior? Because it certainly doesn't look good.

```
[E 20:39:53.187 LabApp] {
      "Host": "127.0.0.1:8888",
      "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:83.0) Gecko/20100101 Firefox/83.0",
      "Accept": "*/*",
      "Accept-Language": "en-US,en;q=0.5",
      "Accept-Encoding": "gzip, deflate, br",
      "Referer": "https://127.0.0.1:8888/lab?",
      "Content-Type": "text/plain;charset=UTF-8",
      "X-Xsrftoken": "2|0d88b5b9|xxx|1607891023",
      "Origin": "https://127.0.0.1:8888",
      "Content-Length": "24",
      "Connection": "keep-alive",
      "Cookie": "username-127-0-0-1-8888=\"2|1:0|10:1607891028|23:username-127-0-0-1-8888|44:xxx=|xxxx\"; _xsrf=2|0d88b5b9|xxx|1607891023",
      "Pragma": "no-cache",
      "Cache-Control": "no-cache"
    }

[E 20:39:53.188 LabApp] 500 POST /git/show_top_level?1607891993159 (127.0.0.1) 23.65ms referer=https://127.0.0.1:8888/lab?
```

I mean, if the directory structure is not a git repository, surely this shouldn't be handled as a 500 error showing up and 'polluting' the terminal output as such? I may be ignorant on this point, but until now I haven't ever seen 'normal' behaviour handled as such in the Jupyter ecosystem.

**Comment by fcollonval** [comment]
Indeed it could be improved by handling that on the server side to avoid those server errors.

So reopening to keep track as an improvement.

**Comment by fcollonval** [comment]
> 500 error appears in the terminal (without debug mode), with an output of the request that caused the 500 error. Is this also part of normal behavior?

This is the behavior of server. Log message with level info or higher (here it is an error as indicated by the `E`) are displayed. This is configurable see https://github.com/jupyter/notebook/blob/0c83c9d59f5ba22b3a09b7db4ebcf3af881ffc53/notebook/notebookapp.py#L748

**Comment by jeanmonet** [comment]
I haven't looked at jupyterlab-git code before, but I think it could be handleded in the following block:
https://github.com/jupyterlab/jupyterlab-git/blob/e3cae9efcdeb3e14e1b994832b9118ee2f00af59/jupyterlab_git/handlers.py#L77

Intuitively I would think a special error handling case should be included when error code is `128`.

**Comment by fcollonval** [comment]
Thanks for digging up.

Unfortunately this is gonna be a bit more complicated than that as git return code is 128 for all errors it encounters :disappointed:. So the check needs to be done on the error message to emit error when something else than `Not in a git repository` occurs.

An example of such handling can be seen in:

https://github.com/jupyterlab/jupyterlab-git/blob/e3cae9efcdeb3e14e1b994832b9118ee2f00af59/jupyterlab_git/git.py#L1136

Feel free to propose a PR, I help you with it.

**Comment by jeanmonet** [comment]
Sure, I'll plunge into it in the following days and report back with a PR!

**Comment by jeanmonet** [comment]
sorry was a bit late, but here goes: https://github.com/jeanmonet/jupyterlab-git/pull/1

**Comment by fcollonval** [comment]
> sorry was a bit late, but here goes [PR](https://github.com/jeanmonet/jupyterlab-git/pull/1) proposal.

Don't apologize there is nothing as late when someone is helping :wink: 

One remark though, could you propose the changes against this repository? You are currently pushing your proposal on your fork master and not this repository master.

You must select `jupyterlab/jupyterlab-git` as base repository when initiating the PR (see the [documentation](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork)):

![image](https://user-images.githubusercontent.com/8435071/103866682-7022af00-50c6-11eb-8da9-dfce6e7d8c56.png)

**Comment by jeanmonet** [comment]
You are right, I missed that step when creating the PR. 

I was just reviewing my proposal, and I think it is not correct to comment line 78:

```python
# self.set_status(500)  # already handled in GitShowTopLevelHandler class
```

If you agree, and before I adjust the PR, the following code:

```python

        if result["code"] != 0:
            # Error messages from git to handle as log.info instead of 500 error
            error_message = map(
                lambda n: n.lower(),
                [
                    # Avoid 500 error messages in terminal when cwd is not a git repo
                    "fatal: not a git repository (or any of the parent directories):",
                ]
            )
            lower_error = result.get("message", "").lower()
            if any([msg in lower_error for msg in error_message]):
                self.log.info(
                    "[jupyterlab_git] Failed to execute 'git' command: {!s},"\
                        " with message: {!s}".format(
                            result["command"], result["message"].strip()
                        )
                )
            else:
                self.set_status(500)
```
should be duplicated for both `GitAllHistoryHandler` and `GitShowTopLevelHandler` classes, and in this case maybe it is better to add the error-handling code in a new method in the `GitHandler` base class which will be called by the subclasses?

**Comment by fcollonval** [comment]
Solved by #855
